proposals/nnnn-actors.md

```diff
@@ -140,6 +140,265 @@ extension BankAccount {
 }
 ```
 
+
+
+### Actor reentrancy (discussion)
+
+One critical point that needs to be discussed and fleshed out is whether actors are [reentrant](https://en.wikipedia.org/wiki/Reentrancy_(computing)) by default or not.
+
+The notion of reentrancy allows the actor runtime to claim the complete elimination of deadlocks, offers opportunity for scheduling optimization techniques where a "high priority task" _must_ be executed as soon as possible, however at the cost of _interleaving_ with any other actor-isolated function's execution any other asynchronous function declared on this actor. This imposes a large mental burden on developers, as every single actor function currently has to be implemented keeping reentrancy in mind, which effectively fails on delivering the promise of pain-free concurrency for the majority of use-cases.
+
+Currently, this proposal takes the aggressive approach of assuming _all_ actors are reentrant and not providing ways to opt out of this behavior. This section aims to highlight issues, benefits and tradeoffs with this approach as well as non-reentrancy in general. The goal of discussing these reentrancy issues, is to arrive at a design that is developer friendly, non-surprising, and delivers more completely on Swift's promise of _safe_ and _sane_ concurrency by default thanks to the use of actors.
+
+#### Reentrant actors: "Interleaving" execution
+
+Reentrancy means that execution of asynchronous actor-isolated functions may "interleave" at suspension points, leading to increased complexity in programming with such actors, as every suspension point must be carefully inspected if the code _after_ it depends on some invariants that could have changed before it suspended.
+
+Interleaving executions still respect the actor's "single-threaded illusion"–i.e. no two functions will ever execute *concurrently* on any given actor–however they may _interleave_ at suspension points. In broad terms this means that reentrant actors are "thread-safe" but are not "data race safe," meaning it is not possible to observe any
```

Comment:
Unfinished thought

Reply:
Also I'm not sure I like calling this a "data race".

Reply:
Hm yeah data race sounds indeed a bit more like the low level one... Perhaps "high-level race free" to somehow point at that it's about high level invariants rather low level synchonization issues?

Reply:
reworded as:

> In broad terms this means that reentrant actors are _thread-safe_ but are automatically protecting from the "high level" kinds of races that may still occur, potentially invalidating invariants upon which an executing asynchronous function may be relying on.